ENH: Add isocalendar accessor to DatetimeIndex and Series.dt

mgmarino opened the pull request:
This PR adds the the isocalendar property to `DatetimeIndex` and the corresponding `Series.dt` accessor.    The property returns a DataFrame, e.g.:

```python
>>> idx = pd.date_range(start='2019-12-29', freq='D', periods=4)
>>> idx.isocalendar
   year  week  day
0  2019    52    7
1  2020     1    1
2  2020     1    2
3  2020     1    3
```

and 

```python
>>> pandas.to_datetime(pandas.Series(["2020-01-01"])).dt.isocalendar
   year  week  day
0  2020     1    
>>> pandas.to_datetime(pandas.Series(["2019-12-31"])).dt.isocalendar.week
0    1
Name: week, dtype: int32
```

The behavior is consistent with `Timestamp.isocalendar` and `datetime.date.isocalendar`.

For more information about ISO 8601 calendar, see e.g. https://en.wikipedia.org/wiki/ISO_week_date.

Address GH33206

- [x] closes #33206 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Note that I am very happy to rename the field, but I wanted to go ahead and open up the PR.

mgmarino committed:
Add function to get_iso_calendar

- This function reproduces what `datetime.date.isocalendar` returns
- Refactor get_week_of_year to use get_iso_calendar internally

WillAyd commented:
Thanks for the PR. I agree the naming is a little weird, though do ew need a dedicated property for this? Might be more natural to just convert to the appropriate period and still use the .year attribute therefrom

@jbrockmendel @mroeschke

mroeschke commented:
Agreed with @WillAyd.

Since `Timestamp` has an `isocalendar` method from the standard library, I wouldn't be opposed to a `DatetimeIndex.isocalendar` or `Series.dt.isocalendar` method that converts the dates to returns a 3-tuple, (ISO year, ISO week number, ISO weekday).

mgmarino commented:
@WillAyd @mroeschke Thanks for the comments!  

I agree that the `isocalendar` method approach is cleaner and more symmetric with `Timestamp`, this was slightly easier for me to get my feet wet with the code.  An additional advantage of this is that it makes it clearer that all values are calculated using ISO (8601).  

Regarding whether this functionality is needed at all, I think if `weekofyear` is provided (which is currently the ISO week), then there absolutely should to be a way to get the year that corresponds to it.

mroeschke commented:
I'd say if we include `isocalendar` (and think of a nice API to easily get the `.year`, `.week`, etc), we could deprecate `weekofyear` (which `datetime.datetime` doesn't have).

Generally, I believe we should align all `.dt` attributes/methods to what exists in `datetime.datetime`)

mgmarino commented:
Yes, I think a series of tuples is not completely user friendly.  What about like what's done in `.dt.components` [here](https://github.com/pandas-dev/pandas/blob/master/pandas/core/indexes/accessors.py#L249), i.e. it returns a frame?  

`isocalendar` could be a property returning a frame (with three columns `year`, `week`, `day`), and then the respective series could simply be accessed like: 

```python
.dt.isocalendar.year
.dt.isocalendar.week
.dt.isocalendar.day
```
What do you think?

Re the deprecation of `weekofyear`, I think that would be a good idea.  I get the impression that there's confusion around it by looking through old issues.  By access `week` through `isocalendar`, it would I think reduce confusion.

mroeschke commented:
+1 to that API @mgmarino. Interested in working on both the `isocalendar` addition and deprecation of  `weekofyear` (and if there's additional attributes like this)?

mgmarino commented:
Definitely!  I'll need a bit of a hand with the Deprecation process, but I think I'm pretty clear how to implement `isocalendar`.  That should probably be done in a separate PR anyways.  

For the `isocalendar` changes, shall I just keep this PR (to have the thread), with an appropriate rename?

mroeschke commented:
> The deprecation should probably be done in a separate PR anyways.

Sure sounds good.

> For the isocalendar changes, shall I just keep this PR (to have the thread), with an appropriate rename?

Definitely. Feel free to retool this PR (and original issue) about adding `isocalendar`

mgmarino committed:
Add fields function to build isocalendar array

- Returns a structured numpy array with year, week, and day
corresponding to the ISO 8601 calendar

mgmarino committed:
Add isocalendar property to DatetimeArray

- Add corresponding access properties in accessors
- This calculates the year, week, and day components according to the ISO 8601 calendar and returns them in a data frame.
- It is analogous to Timestamp.isocalendar and datetime.date.isocalendar

mgmarino committed:
Fix dt tests to handle DataFrame output

mgmarino committed:
Update what’s new for isocalendar

mroeschke commented on pandas/core/arrays/datetimes.py:
```diff
@@ -1234,6 +1234,48 @@ def date(self):
+        iso_calendar_array = self._maybe_mask_results(
+            sarray, fill_value=np.nan, convert=[(n, "<f8") for n in sarray.dtype.names]
+        )
+        return DataFrame(iso_calendar_array)
```
The columns will need to labeled here `DataFrame(iso_calendar_array, columns=['year', 'week', 'day'])`

mgmarino replied:
Thanks, done.

mgmarino committed:
Add columns explicitly when creating DataFrame

mgmarino committed:
Set fill value to a tuple of nan

- Older versions of numpy otherwise throw an error:
```
TypeError: a bytes-like object is required, not 'float'
```

mgmarino commented:
FYI, the failing tests seem to be due to visual studio not liking the ctuple return. I’ll still need some time to investigate a fix and/or rewrite to avoid the ctuple.

mgmarino committed:
Define return of iso_calendar as ctypedef

mgmarino commented:
@mroeschke  

Ok, I'm pretty confident that the build will now succeed (it succeeded on my local Azure pipelines), but here's the news:

- The error was due to a bug in cython fixed in 0.29.16 (See cython/cython#2745, cython/cython#1427)
- Because of this, I had to bump the minimum version of cython to 0.29.16.  

So, I would understand if this feature is not important enough to do this, so I'm happy to have your feedback, if I should:

- leave as is,
- bump the dependency version upgrade out to a separate PR.
- rework the internal code to avoid using ctuple.
- other?

Thanks for your help on this!

mroeschke commented on pandas/tests/series/test_datetime_values.py:
```diff
@@ -665,3 +667,23 @@ def test_setitem_with_different_tz(self):
+        ],
+    )
+    def test_isocalendar(self, input_series, expected_output, expected_type):
+        ser = pd.Series(input_series, dtype="datetime64[D]")
```
You can use `to_datetime` instead of `dtype="datetime64[D]"` here. (We don't actually support `"datetime64[D]"`)

mgmarino replied:
Thanks!

mroeschke commented:
I think it's okay to bump Cython since it's a build dependency cc @jbrockmendel.

mgmarino committed:
Use to_datetime instead of dtype=“datetime64[D]”

mgmarino commented:
Current comments addressed, failing doc build seems to be handled with #33309.

jbrockmendel commented:
i think we're OK to bump cython to the newest bugfix release

mroeschke commented on pandas/core/indexes/accessors.py:
```diff
@@ -219,6 +219,40 @@ def to_pydatetime(self) -> np.ndarray:
+
+        Examples
+        --------
+        >>> pd.to_datetime(pd.Series(["2020-01-01"])).dt.isocalendar
```
I'd just create one datetime `Series` in the beginning and just reference that in the other examples:

```
>>> ser = pd.Series(pd.to_datetime(["2020-01-01", "NaT"]))
>>> ser
...

>>>  ser.dt.isocalendar
>>>  ser.dt.isocalendar.week
```

mroeschke commented on ci/deps/azure-37-numpydev.yaml:
```diff
@@ -15,6 +14,7 @@ dependencies:
   - pytz
   - pip
   - pip:
+    - cython>=0.29.16
```
Why did some of these get moved down to `pip`?

mgmarino replied:
This builds don't use the conda-forge channel and so (for some reason) don't have 0.29.16.  We have to fall back to using pip.

jreback commented on pandas/core/arrays/datetimes.py:
```diff
@@ -1234,6 +1234,50 @@ def date(self):
+
+        sarray = fields.build_isocalendar_sarray(self.asi8)
+        iso_calendar_array = self._maybe_mask_results(
+            sarray,
```
hmm, rather than converting to floats I think we can simply return this as an Int64 frame as this is new code.
so you can put 0 for the nulls and pass a mask

jreback reviewed (changes requested):
can you push the cython update indepdent of this change (and include a whatsnew note for that).

jbrockmendel commented on pandas/_libs/tslibs/ccalendar.pyx:
```diff
@@ -150,33 +150,63 @@ cpdef int32_t get_week_of_year(int year, int month, int day) nogil:
+
+    Returns
+    -------
+    (year : int32_t, week : int32_t, day : int32_t)
```
same 3-line layout as in the Parameters section

jbrockmendel commented on pandas/tests/tslibs/test_ccalendar.py:
```diff
@@ -25,3 +25,27 @@ def test_get_day_of_year_dt():
+    assert (
+        ccalendar.get_iso_calendar(*input_date_tuple)
+        == date(*input_date_tuple).isocalendar()
+    )
```
can you split this into 3 lines
```
result = ...
expected = ...
assert ...
```

mgmarino committed:
Fix doc string to be in row format

mgmarino committed:
Improve readability of ccalendar test

mgmarino committed:
Return Int64 dataframe when NaT present

mgmarino committed:
Clean up example to use a single Series

jreback commented on pandas/tests/series/test_datetime_values.py:
```diff
@@ -665,3 +667,23 @@ def test_setitem_with_different_tz(self):
+        result = pd.to_datetime(pd.Series(input_series)).dt.isocalendar
+        expected_frame = pd.DataFrame(
+            expected_output, columns=["year", "week", "day"]
+        ).astype(expected_type)
```
why do you have to astype here?

I would *always* expect an Int32 type (whether nans or not).

mgmarino replied:
@jreback Ah, I misinterpreted your previous comment, apologies.  Now always returning an Int64.

mgmarino committed:
Always return an Int64 data frame

mgmarino commented:
Rebased, happy to hear any further comments!

jreback commented on pandas/core/arrays/datetimes.py:
```diff
@@ -1234,6 +1234,50 @@ def date(self):
+
+        sarray = fields.build_isocalendar_sarray(self.asi8)
+        iso_calendar_df = DataFrame(
+            sarray, columns=["year", "week", "day"], dtype="Int64"
```
use UInt32 (as that's the default for all calendar component ops)

jreback commented on pandas/_libs/tslibs/fields.pyx:
```diff
@@ -670,3 +670,42 @@ cpdef isleapyear_arr(ndarray years):
+    with nogil:
+        for i in range(count):
+            if dtindex[i] == NPY_NAT:
+                ret_val = -1, -1, -1
```
make these return 0

jreback commented on pandas/core/indexes/accessors.py:
```diff
@@ -219,6 +219,38 @@ def to_pydatetime(self) -> np.ndarray:
+        >>> ser.dt.isocalendar.week
+        0      53
+        1    <NA>
+        Name: week, dtype: Int64
```
you will need to update this as per the dtype change

jreback reviewed (changes requested):
some comments. can you add some prose in timeseries.rst about this (basically just showing it, add a versionadded 1.1 tag). ping on green.

mgmarino committed:
Return UInt32 per calendar ops standard

mgmarino committed:
Add timeseries documentation

mgmarino commented:
@jreback Thanks for the comments, I finally understood it now.   

Dtype -> UInt32, Doc added, Tests green.

jreback commented:
lgtm. cc @TomAugspurger ok here?

jreback committed:
Merge branch 'master' into add-year-for-week-of-year

jreback commented:
this lgtm. let's merge on green.

mroeschke commented:
Thanks @mgmarino! Very nice. Happy to have a follow up to deprecate weekofyear.

jorisvandenbossche commented:
Since `isocalendar`is a *method* and not an attribute on `Timestamp`, should we make `dt.isocalendar` a method instead of a property as well for consistency?

mroeschke commented:
Good point @jorisvandenbossche. 

@mgmarino mind creating a follow up PR that makes `isocalendar` a method instead of an attribute?

mgmarino commented:
> mind creating a follow up PR that makes isocalendar a method instead of an attribute?

@mroeschke Sure no problem, seems reasonable.  The PR is open, I'll ping on green.  

> Happy to have a follow up to deprecate weekofyear.

Will do soon!
